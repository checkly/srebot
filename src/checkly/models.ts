export type ErrorMessage = {
  message: string;
  stack: string;
};
export type LogEntry = {
  time: number;
  msg: string;
  level: string;
};

export class CheckGroup {
  id: number;
  name: string;
  concurrency: number;
  apiCheckDefaults: {
    url: string;
    headers: {
      key: string;
      value: string;
      locked: boolean;
    }[];
    basicAuth: {
      password: string;
      username: string;
    };
    assertions: any[];
    queryParameters: any[];
  };
  alertSettings: {
    reminders: {
      amount: number;
      interval: number;
    };
    escalationType: string;
    runBasedEscalation: {
      failedRunThreshold: number;
    };
    timeBasedEscalation: {
      minutesFailingThreshold: number;
    };
  };
  environmentVariables: any[];
  setupSnippetId: string | null;
  tearDownSnippetId: string | null;
  localSetupScript: string | null;
  localTearDownScript: string | null;
  activated: boolean;
  muted: boolean;
  useGlobalAlertSettings: boolean;
  doubleCheck: boolean;
  locations: string[];
  tags: any[];
  created_at: string;
  updated_at: string;
  runtimeId: string | null;
  retryStrategy: string | null;
  runParallel: boolean;
  alertChannelSubscriptions: {
    alertChannelId: number;
    activated: boolean;
  }[];
  privateLocations: any[];
}

export class Check {
  id: string;
  checkType: string;
  name: string;
  frequency: number;
  frequencyOffset: number;
  activated: boolean;
  muted: boolean;
  shouldFail: boolean;
  locations: string[];
  script: string | null;
  created_at: string;
  updated_at: string;
  environmentVariables: any[];
  doubleCheck: boolean;
  tags: any[];
  sslCheckDomain: string;
  setupSnippetId: string | null;
  tearDownSnippetId: string | null;
  localSetupScript: string | null;
  localTearDownScript: string | null;
  alertSettings: {
    reminders: {
      amount: number;
      interval: number;
    };
    escalationType: string;
    runBasedEscalation: {
      failedRunThreshold: number;
    };
    timeBasedEscalation: {
      minutesFailingThreshold: number;
    };
  };
  useGlobalAlertSettings: boolean;
  degradedResponseTime: number;
  maxResponseTime: number;
  groupId: number | null;
  groupOrder: number | null;
  runtimeId: string | null;
  scriptPath: string | null;
  retryStrategy: string | null;
  runParallel: boolean;
  request: {
    method: string;
    url: string;
    body: string;
    bodyType: string;
    headers: {
      key: string;
      value: string;
      locked: boolean;
    }[];
    queryParameters: any[];
    assertions: any[];
    basicAuth: {
      password: string;
      username: string;
    };
    followRedirects: boolean;
    skipSSL: boolean;
    ipFamily: string;
  };
  alertChannelSubscriptions: {
    alertChannelId: number;
    activated: boolean;
  }[];
  privateLocations: any[];
}

export class CheckResult {
  id: string;
  name: string;
  checkId: string;
  hasFailures: boolean;
  hasErrors: boolean;
  isDegraded: boolean;
  overMaxResponseTime: boolean;
  runLocation: string;
  startedAt: string; // ISO date string
  stoppedAt: string; // ISO date string
  created_at: string; // ISO date string
  responseTime: number;
  apiCheckResult: {
    assertions: Array<{
      source: string;
      target: number;
    }>;
    request: {
      method: string;
      url: string;
      data: string;
      headers: Record<string, string>;
      params: Record<string, string>;
    };
    response: {
      status: number;
      statusText: string;
      body: string;
      headers: Record<string, string>;
      timings: Record<string, unknown>;
      timingPhases: Record<string, unknown>;
    };
    requestError: string | null;
    jobLog: {
      setup: Array<LogEntry>;
      request: Array<LogEntry>;
      teardown: Array<LogEntry>;
    };
    jobAssets: unknown | null;
  } | null;
  browserCheckResult: {
    jobLog: Array<LogEntry>;
    playwrightTestTraces: Array<string>;
  } | null;
  multiStepCheckResult: {
    errors: Array<ErrorMessage>;
    endTime: number;
    startTime: number;
    runtimeVersion: string;
    jobLog: Array<LogEntry>;
    jobAssets: Array<unknown>;
    playwrightTestTraces: Array<string>;
    playwrightTestJsonReportFile: string;
  } | null;
  checkRunId: number;
  attempts: number;
  resultType: string;
  sequenceId: string;

  getLog(): string {
    const jobLog =
      this.apiCheckResult?.jobLog.request ||
      this.browserCheckResult?.jobLog ||
      this.multiStepCheckResult?.jobLog ||
      [];

    return jobLog
      .map(
        (logEntry) => `${logEntry.time} - ${logEntry.level}: ${logEntry.msg}`
      )
      .join("\n");
  }
}
