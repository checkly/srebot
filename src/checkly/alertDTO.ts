import { Transform } from 'class-transformer';
import {
  IsString,
  IsUUID,
  IsNumber,
  IsArray,
  IsOptional,
} from 'class-validator';

export class AlertDto {

  @IsString()
  CHECK_NAME: string;

  @IsUUID()
  CHECK_ID: string;
  @IsUUID()
  $UUID: string;

  @IsString()
  CHECK_TYPE: string;

  @IsString()
  GROUP_NAME: string;

  @IsString()
  ALERT_TITLE: string;

  @IsString()
  ALERT_TYPE: string;

  @IsUUID()
  CHECK_RESULT_ID: string;

  @IsNumber()
  RESPONSE_TIME: number;

  @IsOptional() // This is optional because it's only for API checks
  @IsNumber()
  API_CHECK_RESPONSE_STATUS_CODE?: number;

  @IsOptional() // This is optional because it's only for API checks
  @IsString()
  API_CHECK_RESPONSE_STATUS_TEXT?: string;

  @IsString()
  RUN_LOCATION: string;

  @IsString()
  RESULT_LINK: string;

  @IsOptional() // This is only for ALERT_SSL alerts
  @IsNumber()
  SSL_DAYS_REMAINING?: number;

  @IsOptional() // This is only for ALERT_SSL alerts
  @IsString()
  SSL_CHECK_DOMAIN?: string;

  @IsString()
  STARTED_AT: string;

  @Transform(({ value }) => {
    try {
      if (!value) {
        return [];
      }
      console.log('jsonarray:' + value);
      // If the value is a valid stringified JSON array, parse it
      const parsed = typeof value === 'string' ? JSON.parse(value) : value;

      // Return the value only if it's a valid array, otherwise return an empty array
      return Array.isArray(parsed) ? parsed : [];
    } catch (e) {
      if (e instanceof SyntaxError) {
        return [value];
      }
      // If parsing fails, return an empty array
      console.trace(e);
      return [];
    }
  })
  @IsArray() // Assuming TAGS is an array of strings
  @IsString({ each: true })
  TAGS: string[];

  @IsNumber()
  $RANDOM_NUMBER: number;

  @IsString()
  moment: string;

  constructor(json: Partial<AlertDto>) {
    this.CHECK_NAME = json.CHECK_NAME || '';
    this.CHECK_ID = json.CHECK_ID || '';
    this.$UUID = json.$UUID || '';
    this.CHECK_TYPE = json.CHECK_TYPE || '';
    this.GROUP_NAME = json.GROUP_NAME || '';
    this.ALERT_TITLE = json.ALERT_TITLE || '';
    this.ALERT_TYPE = json.ALERT_TYPE || '';
    this.CHECK_RESULT_ID = json.CHECK_RESULT_ID || '';
    this.RESPONSE_TIME = json.RESPONSE_TIME ? Number(json.RESPONSE_TIME) : 0;
    this.API_CHECK_RESPONSE_STATUS_CODE = json.API_CHECK_RESPONSE_STATUS_CODE || 0;
    this.API_CHECK_RESPONSE_STATUS_TEXT = json.API_CHECK_RESPONSE_STATUS_TEXT || '';
    this.RUN_LOCATION = json.RUN_LOCATION || '';
    this.RESULT_LINK = json.RESULT_LINK || '';
    this.SSL_DAYS_REMAINING = json.SSL_DAYS_REMAINING || 0;
    this.SSL_CHECK_DOMAIN = json.SSL_CHECK_DOMAIN || '';
    this.STARTED_AT = json.STARTED_AT || '';
    this.TAGS = json.TAGS || [];
    this.$RANDOM_NUMBER = json.$RANDOM_NUMBER ? Number(json.$RANDOM_NUMBER) : 0;
    this.moment = json.moment || '';
  }

}
