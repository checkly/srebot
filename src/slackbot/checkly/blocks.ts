import {
  CHECKLY_ANALYZE_CHECK_ACTION,
  CHECKLY_LIST_CHECKS_ACTIONS,
  CHECKLY_RUN_CHECK_ACTION,
} from "./slack";

export const summaryBlock = (
  failed: number,
  degraded: number,
  passed: number,
) => {
  return {
    text: `Checkly Production Status: ${failed} Failed | ${degraded} Degraded | ${passed} Passed`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: "Checkly Production Status",
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `:x: *${failed}* Failed  |  :warning: *${degraded}* Degraded  |  :white_check_mark: *${passed}* Passed\n`,
        },
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Failing Checks",
              emoji: true,
            },
            action_id: CHECKLY_LIST_CHECKS_ACTIONS.FAILING,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Degraded Checks",
              emoji: true,
            },
            action_id: CHECKLY_LIST_CHECKS_ACTIONS.DEGRADED,
          },
        ],
      },
    ],
  };
};

export const checkAttachment = ({
  checkId,
  check,
  group,
  type,
  tags,
  success24h,
  success7d,
  timestamp,
  location,
  executionTime,
  executionTimeAvg,
  executionTimeP95,
  executionTimeP99,
}: {
  checkId: string;
  check: string;
  group: string;
  type: string;
  tags: string[];
  success24h: number;
  success7d: number;
  timestamp: Date;
  location: string;
  executionTime: number;
  executionTimeAvg: number;
  executionTimeP95: number;
  executionTimeP99: number;
}) => {
  return {
    blocks: [
      {
        type: "divider",
      },
      {
        type: "header",
        text: {
          type: "plain_text",
          text: check,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Type:* ${type}\n*Group:* ${group}\n*Tags*: ${tags.join(", ")}`,
        },
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*When*\n<!date^${Math.floor(timestamp.getTime() / 1000)}^{ago}|${timestamp}>`,
          },
          {
            type: "mrkdwn",
            text: "*Region*\n" + location,
          },
        ],
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*Execution Time*\n${executionTime}s (${executionTimeAvg}s Avg | ${executionTimeP95}s P95 | ${executionTimeP99}s P99)`,
          },
          {
            type: "mrkdwn",
            text: `*Reliability*\n${success24h}% in 24h | ${success7d}% in 7d`,
          },
        ],
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "View Check",
              emoji: true,
            },
            url: `https://app.checklyhq.com/checks/${checkId}`,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Run Check",
              emoji: true,
            },
            value: checkId,
            action_id: CHECKLY_RUN_CHECK_ACTION,
          },
          {
            type: "button",
            text: {
              type: "plain_text",
              text: "Analyze",
              emoji: true,
            },
            value: checkId,
            action_id: CHECKLY_ANALYZE_CHECK_ACTION,
          },
        ],
      },
    ],
  };
};
