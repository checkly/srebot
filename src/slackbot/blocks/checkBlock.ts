import { Check, CheckResult } from "../../checkly/models";
import { SummarizeErrorsPromptType } from "../../prompts/checkly";

export interface CheckBlockProps {
  check: Check;
  failureCount: number;
  errorGroups?: SummarizeErrorsPromptType;
  checkResults: CheckResult[];
}

export function createCheckBlock({
  check,
  failureCount,
  errorGroups = { groups: [] },
  checkResults,
}: CheckBlockProps) {
  return {
    text: `*Check Details*`,
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: `${check.name} - Last 24 hours`,
          emoji: true,
        },
      },
      {
        type: "divider",
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*Type*\n${
              {
                BROWSER: "Browser Check",
                API: "API Check",
                MULTI_STEP: "Multi-Step Check",
              }[check.checkType]
            }`,
          },
          {
            type: "mrkdwn",
            text: `*Frequency*\nevery *${check.frequency}* minute${check.frequency > 1 ? "s" : ""}`,
          },
        ],
      },
      {
        type: "section",
        fields: [
          {
            type: "mrkdwn",
            text: `*Locations*\n\`${check.locations.join("\`, \`")}\``,
          },
          {
            type: "mrkdwn",
            text: `*Failure Rate*\n${((failureCount / checkResults.length) * 100).toFixed(2).replace(/\.00$/, "")}% (${failureCount} / ${checkResults.length})`,
          },
        ],
      },
      ...(errorGroups.groups.length === 0
        ? [
            {
              type: "section",
              text: {
                type: "mrkdwn",
                text: `*No errors happened in the last 24 hours*`,
              },
            },
          ]
        : [
            {
              type: "header",
              text: {
                type: "plain_text",
                text: `${errorGroups.groups.length > 0 ? "Detected" : "No"} Error Patterns`,
                emoji: true,
              },
            },
          ]),
      ...errorGroups.groups.flatMap((group) => [
        {
          type: "divider",
        },
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: `*Pattern:* \`${group.errorMessage}\``,
          },
        },
        {
          type: "section",
          fields: [
            {
              type: "mrkdwn",
              text: `*Count*\n*${group.checkResults.length}* failure${group.checkResults.length > 1 ? "s" : ""}`,
            },
            {
              type: "mrkdwn",
              text: `*Affected Locations*\n\`${[
                ...new Set(
                  group.checkResults.map(
                    (id) => checkResults.find((r) => r.id === id)?.runLocation,
                  ),
                ),
              ]
                .filter(Boolean)
                .join("\`, \`")}\``,
            },
          ],
        },
      ]),
    ],
  };
}
