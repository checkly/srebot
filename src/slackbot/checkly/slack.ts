import { PrometheusCheckMetric, Reporting, Status } from "../../checkly/models";
import { checkly } from "../../checkly/client";
import { checkAttachment, summaryBlock } from "./blocks";
import { featureCoveragePrompt } from "../../prompts/checkly";
import { generateText } from "ai";
import { text } from "express";

export const CHECKLY_COMMAN_NAME = "/checkly";

export const checklyCommandHandler = async ({ ack, respond }) => {
  await ack();

  const metrics = await checkly.getPrometheusCheckStatus();

  const failing = metrics.failing.length;
  const degraded = metrics.degraded.length;
  const passing = metrics.passing.length;

  const { text, blocks } = summaryBlock(failing, degraded, passing);

  await respond({
    text,
    blocks,
  });
};

export const CHECKLY_LIST_CHECKS_ACTION =
  /checkly-list-checks-(failing|degraded|passing)/;

export enum CHECKLY_LIST_CHECKS_ACTIONS {
  FAILING = "checkly-list-checks-failing",
  DEGRADED = "checkly-list-checks-degraded",
  PASSING = "checkly-list-checks-passing",
}

export const checklyListChecksActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  const [rep24h, rep7d, statuses, metrics] = await Promise.all([
    checkly.getReporting({
      quickRange: "last24Hrs",
    }),
    checkly.getReporting({ quickRange: "last7Days" }),
    checkly.getStatuses(),
    checkly.getPrometheusCheckStatus(),
  ]);

  const rep24hMap = rep24h.reduce(
    reduceReporting,
    {} as Record<string, Reporting>,
  );
  const rep7dMap = rep7d.reduce(
    reduceReporting,
    {} as Record<string, Reporting>,
  );

  const statusesMap = statuses.reduce(
    reduceStatus,
    {} as Record<string, Status>,
  );

  const filter = action.action_id as CHECKLY_LIST_CHECKS_ACTIONS;
  let checksToShow: any[];
  let attachmentColor: string;

  switch (filter) {
    case CHECKLY_LIST_CHECKS_ACTIONS.FAILING:
      checksToShow = metrics.failing;
      attachmentColor = "#ff0000"; // Fixed syntax error - changed : to =
      break;
    case CHECKLY_LIST_CHECKS_ACTIONS.DEGRADED:
      checksToShow = metrics.degraded;
      attachmentColor = "#ffa500";
      break;
    case CHECKLY_LIST_CHECKS_ACTIONS.PASSING:
      checksToShow = metrics.passing;
      attachmentColor = "#36a64f";
      break;
    default:
      checksToShow = [];
      attachmentColor = "";
  }

  const attachments = checksToShow
    .map(PrometheusCheckMetric.fromJson)
    .map((m) => ({
      checkRunId: statusesMap[m.checkId].lastCheckRunId,
      checkId: m.checkId,
      check: m.name,
      group: m.group,
      type: m.checkType,
      tags: m.tags,
      success24h: rep24hMap[m.checkId].aggregate.successRatio,
      success7d: rep7dMap[m.checkId].aggregate.successRatio,
      timestamp: new Date(statusesMap[m.checkId].updated_at),
      location: statusesMap[m.checkId].lastRunLocation || "Unknown",
      executionTime: rep7dMap[m.checkId].aggregate.avg,
      executionTimeAvg: rep7dMap[m.checkId].aggregate.avg,
      executionTimeP95: rep7dMap[m.checkId].aggregate.p95,
      executionTimeP99: rep7dMap[m.checkId].aggregate.p99,
    }))
    .map((m) => ({
      color: attachmentColor,
      blocks: checkAttachment(m).blocks,
    }));

  await respond({
    text: "Checks:",
    attachments,
  });

  return;
};

const reduceReporting = (
  acc: Record<string, Reporting>,
  curr: Reporting,
): Record<string, Reporting> => {
  acc[curr.checkId] = curr;
  return acc;
};

const reduceStatus = (
  acc: Record<string, Status>,
  curr: Status,
): Record<string, Status> => {
  acc[curr.checkId] = curr;
  return acc;
};

export const CHECKLY_RUN_CHECK_ACTION = "checkly-run-check";

export const checklyRunCheckActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  await checkly.runCheck(JSON.parse((action as any).value).checkId);

  await respond("Check triggered!");
};

export const CHECKLY_ANALYZE_CHECK_ACTION = "checkly-analyze-check";

export const checklyAnalyzeCheckActionHandler = async ({
  action,
  ack,
  respond,
}) => {
  await ack();

  const checkId = action.value;

  const check = await checkly.getCheck(checkId);
  const lastFailingCheckResult = await checkly.getCheckResults(
    checkId,
    true,
    1,
  );

  if (lastFailingCheckResult.length == 0) {
    await respond(
      `Unable to find failure for check ${checkId}.\n<https://app.checklyhq.com/checks/${checkId}|Open the check in your Browser instead.>`,
    );
    return;
  }

  if (!lastFailingCheckResult[0].browserCheckResult) {
    await respond(
      `Check ${checkId} is not a Browser check.\n<https://app.checklyhq.com/checks/${checkId}|Open the check in your Browser instead.>`,
    );
    return;
  }

  const errors = lastFailingCheckResult[0].browserCheckResult!.errors.map(
    (e) => e.stack,
  );

  const [prompt, config] = featureCoveragePrompt(
    check.name,
    check.scriptPath || "unknown",
    check.script || "unknown",
    errors,
  );
  const { text: summary } = await generateText({
    ...config,
    prompt,
  });

  await respond({
    text: summary,
  });
};
