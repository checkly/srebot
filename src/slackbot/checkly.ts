import { generateObject } from "ai";
import { checkly } from "../checkly/client";
import {
  summarizeErrorsPrompt,
  SummarizeErrorsPromptType,
} from "../prompts/checkly";
import {
  fetchCheckResults,
  last24h,
  LAST_30_DAYS,
  summarizeCheckResult,
} from "../prompts/checkly-data";
import { createCheckResultBlock } from "./blocks/checkResultBlock";
import { generateHeatmapPNG } from "../heatmap/createHeatmap";
import { createCheckBlock } from "./blocks/checkBlock";

async function checkResultSummary(checkId: string, checkResultId: string) {
  const check = await checkly.getCheck(checkId);
  const checkAppUrl = checkly.getCheckUrl(check.id);
  const checkResult = await checkly.getCheckResult(check.id, checkResultId);
  const checkResultAppUrl = checkly.getCheckResultAppUrl(
    check.id,
    checkResult.id,
  );

  const interval = last24h(new Date(checkResult.startedAt));

  const checkResults = await fetchCheckResults(checkly, {
    checkId: check.id,
    ...interval,
  });

  console.log("CHECK_RESULTS", checkResults.length);

  const failingCheckResults = checkResults.filter(
    (result) => result.hasFailures || result.hasErrors,
  );

  console.log("CHECK_RESULT_FAILURES", failingCheckResults.length);

  const promptDef = summarizeErrorsPrompt({
    check: check.id,
    locations: check.locations,
    frequency: check.frequency,
    interval,
    results: [...failingCheckResults, checkResult].map(summarizeCheckResult),
  });
  const { object: errorGroups } =
    await generateObject<SummarizeErrorsPromptType>(promptDef);

  console.log("ERROR_GROUPS", errorGroups);

  return createCheckResultBlock({
    check,
    checkAppUrl,
    checkResult,
    checkResultAppUrl,
    errorGroups,
    failingCheckResults,
    intervalStart: LAST_30_DAYS.from,
  });
}

async function checkSummary(checkId: string) {
  const check = await checkly.getCheck(checkId);

  const interval = last24h(new Date());

  const checkResults = await fetchCheckResults(checkly, {
    checkId: check.id,
    ...interval,
  });

  console.log("CHECK_RESULTS", checkResults.length);
  // console.log("CHECK_RESULTS", JSON.stringify(checkResults, null, 2));

  const failingCheckResults = checkResults.filter(
    (result) => result.hasFailures || result.hasErrors,
  );

  if (failingCheckResults.length === 0) {
    return {
      message: createCheckBlock({
        check,
        checkAppUrl: checkly.getCheckUrl(check.id),
        checkResults,
      }),
      image: null,
    };
  }

  console.log("CHECK_RESULT_FAILURES", failingCheckResults.length);

  const promptDef = summarizeErrorsPrompt({
    check: check.id,
    locations: check.locations,
    frequency: check.frequency,
    interval,
    results: failingCheckResults.map(summarizeCheckResult),
  });
  const { object: errorGroups } =
    await generateObject<SummarizeErrorsPromptType>(promptDef);

  console.log("ERROR_GROUPS", errorGroups);

  const heatmapImage = generateHeatmapPNG(checkResults, {
    bucketSizeInMinutes: check.frequency * 10,
    verticalSeries: check.locations.length,
  });

  const message = createCheckBlock({
    check,
    checkAppUrl: checkly.getCheckUrl(check.id),
    errorGroups,
    checkResults,
  });

  return { message, image: heatmapImage };
}
export const CHECKLY_COMMAND_NAME = "/checkly";

export const checklyCommandHandler = async ({ ack, respond, command }) => {
  await ack();

  const args = command.text.split(" ");
  if (args.length === 1 && !!args[0]) {
    const { message, image } = await checkSummary(args[0]);
    await respond({
      ...message,
    });

    // FIXME find a way to send the image to slack (al)
  } else if (args.length === 2) {
    const [checkId, checkResultId] = args;
    const summary = await checkResultSummary(checkId, checkResultId);

    await respond({
      ...summary,
    });
  } else {
    await respond({
      text: "Please provide either a check ID or both a check ID and check result ID in the format: /checkly <check_id> (<check_result_id>)",
    });
  }
};
