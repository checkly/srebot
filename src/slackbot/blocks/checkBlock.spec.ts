import { describe, expect, it } from "@jest/globals";
import { createCheckBlock } from "./checkBlock";
import { Check, CheckResult } from "../../checkly/models";

describe("checkBlock", () => {
  it("should create a message block for a check summary", () => {
    const mockCheck: Check = {
      id: "test-check-id",
      name: "Test Check",
      checkType: "BROWSER",
      frequency: 10,
      locations: ["us-east-1", "eu-west-1"],
    } as Check;

    const mockCheckResults: CheckResult[] = [
      {
        id: "test-result-id-1",
        runLocation: "us-east-1",
      } as CheckResult,
      {
        id: "test-result-id-2",
        runLocation: "eu-west-1",
      } as CheckResult,
    ];

    const mockErrorGroups = {
      groups: [
        {
          errorMessage: "Test error message",
          checkResults: ["test-result-id-1", "test-result-id-2"],
        },
      ],
    };

    const result = createCheckBlock({
      check: mockCheck,
      checkAppUrl: "https://app.checklyhq.com/checks/test-check-id",
      errorGroups: mockErrorGroups,
      checkResults: mockCheckResults,
    });

    expect(result.blocks).toBeDefined();
    expect(result.blocks[0].type).toBe("header");
    expect(result.blocks[0].text?.text).toContain("Test Check");

    // Verify check details section
    const fields = result.blocks[2]?.fields;
    expect(fields?.[0]?.text).toContain("Browser");
    expect(fields?.[1]?.text).toContain("\`10\` minutes");
    expect(fields?.[2]?.text).toContain("us-east-1");
    expect(fields?.[2]?.text).toContain("eu-west-1");

    // Verify error patterns section
    expect(result.blocks[3].text?.text).toBe("Detected Error Patterns");

    // Verify error group details
    const errorGroupFields = result.blocks[6].fields;
    expect(errorGroupFields?.[0]?.text).toContain("\`2\` failures");
    expect(errorGroupFields?.[1]?.text).toContain("us-east-1");
    expect(errorGroupFields?.[1]?.text).toContain("eu-west-1");
  });
});
