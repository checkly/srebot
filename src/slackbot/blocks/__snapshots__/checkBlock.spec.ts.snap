// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkBlock should create a message block for a check summary 1`] = `
[
  {
    "text": {
      "emoji": true,
      "text": "Test Check - Last 24 hours",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "type": "divider",
  },
  {
    "fields": [
      {
        "text": "*Type:*
Browser",
        "type": "mrkdwn",
      },
      {
        "text": "*Frequency:*
every \`10\` minutes",
        "type": "mrkdwn",
      },
      {
        "text": "*Locations:*
\`us-east-1\`, \`eu-west-1\`",
        "type": "mrkdwn",
      },
      {
        "text": "*Link:*
<https://app.checklyhq.com/checks/test-check-id|Link>",
        "type": "mrkdwn",
      },
    ],
    "type": "section",
  },
  {
    "text": {
      "emoji": true,
      "text": "Detected Error Patterns",
      "type": "plain_text",
    },
    "type": "header",
  },
  {
    "type": "divider",
  },
  {
    "text": {
      "text": "*Pattern:* \`Test error message\`",
      "type": "mrkdwn",
    },
    "type": "section",
  },
  {
    "fields": [
      {
        "text": "*Count:* \`2\` failures",
        "type": "mrkdwn",
      },
      {
        "text": "*Affected Locations:* \`us-east-1\`, \`eu-west-1\`",
        "type": "mrkdwn",
      },
    ],
    "type": "section",
  },
]
`;
